[docs] Expo Module API tutorial (#19801)
expo/expo: docs/pages/modules/existing-library.mdx

---

> title: Intgerating in an existing library

Reviewer: ```suggestion
title: Integrating in an existing library
```

Author: 😅

---

import { CodeBlocksTable } from '~/components/plugins/CodeBlocksTable';
import { Terminal } from '~/ui/components/Snippet';

There are cases where you may want to integrate the Expo Modules API into an existing React Native library. For example, this may be useful to incrementally rewrite the library or to take advantage of [iOS AppDelegate subscribers](/modules/appdelegate-subscribers/) and [Android Lifecycle listeners](/modules/android-lifecycle-listeners/) to automatically set up your library.

The following steps will set up your existing React Native library to access Expo Modules APIs.

Create the [**expo-module.config.json**](./module-config.mdx) file at the root of your project and start from the empty object `{}` inside it. We will fill it in later to enable specific features.

> The presence of the module config is enough for [Expo Autolinking](/workflow/glossary-of-terms/#expo-autolinking) to recognize it as an Expo module and automatically link your native code.

Reviewer: Since we now have a dedicated page for autolinking, let's use it instead.
```suggestion
The presence of the module config is enough for [Expo Autolinking](./autolinking.mdx) to recognize it as an Expo module and automatically link your native code.
```

### Add the `expo-modules-core` native dependency

Add `expo-modules-core` as a dependency in your podspec and **build.gradle** files.<br/>

<CodeBlocksTable tabs={["*.podspec", "build.gradle"]}>

```ruby
# ...
Pod::Spec.new do |s|
  # ...
  s.dependency 'ExpoModulesCore'
end
```

```groovy
// ...
dependencies {
  // ...
  implementation project(':expo-modules-core')
}
```

</CodeBlocksTable>

### Add Expo packages to dependencies

> Add `expo` package as a peer dependency in your **package.json** — we recommend using `*` as a version range so as not to cause any duplicated packages in user's **node_modules** folder. Your library also needs to depend on `expo-modules-core` but only as a dev dependency — it's already provided in the projects depending on your library by the `expo` package with the version of core that is compatible with the specific SDK used in the project.<br/>

Reviewer 2: ```suggestion
Add `expo` package as a peer dependency in your **package.json** — we recommend using `*` as a version range so as not to cause any duplicated packages in user's **node_modules** folder. Your library also needs to depend on `expo-modules-core` but only as a dev dependency — it's already provided in the projects depending on your library by the `expo` package with the version of core that is compatible with the specific SDK used in the project.
```

In general, we should not use `<br />`s since we manage the spacing on the design level. Is there an issue there, or it is just a leftover?

Reviewer: https://github.com/expo/expo/pull/19291#discussion_r983669865
EDIT: This doc is mostly a copy of the previous Get Started page
